test/e2e/ratelimit/ratelimit_universal.go

```diff
@@ -0,0 +1,170 @@
+package ratelimit
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/kumahq/kuma/pkg/config/core"
+	. "github.com/kumahq/kuma/test/framework"
+)
+
+func RateLimitOnUniversal() {
+	var cluster Cluster
+	var deployOptsFuncs []DeployOptionsFunc
+	rateLimitPolicy := `
+type: RateLimit
+mesh: default
+name: rate-limit-all-sources
+sources:
+- match:
+   kuma.io/service: "*"
+destinations:
+- match:
+   kuma.io/service: echo-server_kuma-test_svc_8080
+conf:
+  http:
+    requests: 2
+    interval: 10s
+    onRateLimit:
+      status: 423
+      headers:
+        - key: "x-kuma-rate-limited"
+          value: "true"
+`
+
+	E2EBeforeSuite(func() {
+		clusters, err := NewUniversalClusters(
+			[]string{Kuma3},
+			Silent)
+		Expect(err).ToNot(HaveOccurred())
+
+		// Global
+		cluster = clusters.GetCluster(Kuma3)
+		deployOptsFuncs = KumaUniversalDeployOpts
+
+		err = NewClusterSetup().
+			Install(Kuma(core.Standalone, deployOptsFuncs...)).
+			Setup(cluster)
+		Expect(err).ToNot(HaveOccurred())
+
+		demoClientToken, err := cluster.GetKuma().GenerateDpToken("default", "demo-client")
+		Expect(err).ToNot(HaveOccurred())
+
+		echoServerToken, err := cluster.GetKuma().GenerateDpToken("default", "echo-server_kuma-test_svc_8080")
+		Expect(err).ToNot(HaveOccurred())
+
+		webToken, err := cluster.GetKuma().GenerateDpToken("default", "web")
+		Expect(err).ToNot(HaveOccurred())
+
+		err = YamlUniversal(rateLimitPolicy)(cluster)
+		Expect(err).ToNot(HaveOccurred())
+
+		err = NewClusterSetup().
+			Install(EchoServerUniversal(AppModeEchoServer, "default", "universal", echoServerToken, WithTransparentProxy(true))).
+			Install(DemoClientUniversal(AppModeDemoClient, "default", demoClientToken, WithTransparentProxy(true))).
+			Install(DemoClientUniversal("web", "default", webToken, WithTransparentProxy(true))).
+			Setup(cluster)
+		Expect(err).ToNot(HaveOccurred())
+
+		err = cluster.VerifyKuma()
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	E2EAfterSuite(func() {
+		if ShouldSkipCleanup() {
+			return
+		}
+		err := cluster.DeleteKuma(deployOptsFuncs...)
+		Expect(err).ToNot(HaveOccurred())
+
+		err = cluster.DismissCluster()
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	verifyRateLimit := func(client string, total int) func() int {
+		return func() int {
+			succeeded := 0
+			for i := 0; i < total; i++ {
+				_, _, err := cluster.Exec("", "", client, "curl", "-v", "--fail", "echo-server_kuma-test_svc_8080.mesh")
+				if err == nil {
+					succeeded++
+				}
+			}
+
+			return succeeded
+		}
+	}
+
+	It("should limit to 2 requests per 5 sec", func() {
+		Eventually(verifyRateLimit("demo-client", 5), "60s", "5s").Should(Equal(2))
+		Eventually(verifyRateLimit("demo-client", 5), "30s", "1s").Should(Equal(2))
```

Comment:
Is there a reason for intervals to be different?

Reply:
(also leave a comment that we check a second time to see if the limit is deterministic)